### ISSUE TITLE
[stb_perlin] stb_perlin_ridge_noise3 always produces the same values.

### ISSUE THREAD

**Issue by sigmasoldi3r** [nothings/stb#1139]
Hello, first of all this might have been discussed previously but I'm not finding it.

```cpp
    auto c = stb_perlin_ridge_noise3(1, 2, 3, 2, 0.5f, 1, 6);
    auto d = stb_perlin_ridge_noise3(100, 200, 300, 2, 0.5f, 1, 6);
```

The output is always `0.984375` no matter the coordinates.

What I'm missing?

I've followed the docs:

```cpp
// Fractal Noise:
//
// Three common fractal noise functions are included, which produce
// a wide variety of nice effects depending on the parameters
// provided. Note that each function will call stb_perlin_noise3
// 'octaves' times, so this parameter will affect runtime.
//
// float stb_perlin_ridge_noise3(float x, float y, float z,
//                               float lacunarity, float gain, float offset, int octaves)
//
// float stb_perlin_fbm_noise3(float x, float y, float z,
//                             float lacunarity, float gain, int octaves)
//
// float stb_perlin_turbulence_noise3(float x, float y, float z,
//                                    float lacunarity, float gain, int octaves)
//
// Typical values to start playing with:
//     octaves    =   6     -- number of "octaves" of noise3() to sum
//     lacunarity = ~ 2.0   -- spacing between successive octaves (use exactly 2.0 for wrapping output)
//     gain       =   0.5   -- relative weighting applied to each successive octave
//     offset     =   1.0?  -- used to invert the ridges, may need to be larger, not sure
//
//
```

Not tested against other noise functions.

**Comment by sigmasoldi3r** [comment]
After lowering lacunarity value to ~0.1 values start to change, is that intended?

**Comment by rygorous** [comment]
You give two examples, both of which are exact integer coordinates.

Perlin noise's bandlimitedness construction explicitly makes the noise function always return 0 at integer coordinates. This is fully intentional and part of Perlin's definition. All derived functions therefore also do the same thing.

Setting lacunarity to a non-integer value makes some of the octaves being sampled get sampled at non-integer positions. (Note that setting it to a value <=1 is not intended; the higher-numbered octaves are traditionally meant to be of higher, not lower, frequency.)

In general, when using it for e.g. 2D texture synthesis, the scaling in Perlin's original definition (which stb_perlin follows) is intended to be such that the 2D image is contained in [0,1]^2 or similar, i.e. x and y are both between 0 and 1. This makes the first (lowest-frequency) octave go through exactly one grid cell along each axis over the course of the image, while successive higher-frequency octaves go through 2, 4, 8, etc. cells.

If you are scaled such that you are passing something like integer pixel coordinates (or integers offset by 0.5 or similar) to the noise functions, then aside from the obvious pitfall that is sampling exactly at the positions where noise is defined to be 0 if you pass in integers, even the lowest-frequency octave will be completely different on every sample, which is defeating the purpose of using a multiple-octave construction in the first place.

In short, to use it as intended, you need to scale your coordinates way down; integer steps won't work.

**Comment by sigmasoldi3r** [comment]
Thanks for the explaination! I'll scale the input values as mentioned. 

I was sampling always at integer values.
